rasterio/transform.py

```diff
@@ -61,3 +66,154 @@ def array_bounds(height, width, transform):
     w, n = transform.xoff, transform.yoff
     e, s = transform * (width, height)
     return w, s, e, n
+
+
+def _all_iter(sequence):
+    """Check if all the items in a sequence are iterable."""
+    return all((isinstance(item, collections.Iterable) for item in sequence))
+
+
+def complex_iter(n_pos_args):
```

Comment:
@sgillies @perrygeo I got a little fancy.  This decorator lets `xy()` and similar functions hands a bunch of different input configurations.  The docstring has some examples.

I need to finish handling for `xy(*np.nonzero(), transform)` and figure out how to document in `xy()`.  Switching from `row, col` to `*args` may help with documentation since the function signature isn't entirely truthful.